### ISSUE TITLE
Assert in c code generator

### ISSUE THREAD

**Issue by vegansk** [nim-lang/Nim#5389]
Minimal example:

```nim
import future, sequtils

type
  Validation[A] = object
  Validator[X, A] = X -> Validation[A]

proc mkValidator[A](
  msg: string,
  predicate:  A -> bool
): Validator[A, A] =
  var a: A
  let b = a

block:
  type
    Y = object
      yVal: string
      zObj: Z
    Z = object
      zVal: bool

  let
    valY = mkValidator("yVal is empty", (y: Y) => y.yVal.len > 0)

block:
  type
    Y = object
      yVal: string
      zObj: Z
    Z = object
      zVal: bool

  let
    valY = mkValidator("yVal is empty", (y: Y) => y.yVal.len > 0)
```

fails on that assert: https://github.com/nim-lang/Nim/blob/devel/compiler/ccgexprs.nim#L201

**Comment by vegansk** [comment]
Duplicate of #5170

**Comment by vegansk** [comment]
No, #5170 doesn't fail on that assert

**Comment by vegansk** [comment]
Fixed in 6499462303f1e4ccc8e9a3174985f563e688f566
